internal/api/handler/promote_environment_v1alpha1.go

```diff
@@ -0,0 +1,84 @@
+package handler
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/bufbuild/connect-go"
+	"github.com/pkg/errors"
+	kubeerr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kubev1alpha1 "github.com/akuity/kargo/api/v1alpha1"
+	svcv1alpha1 "github.com/akuity/kargo/pkg/api/service/v1alpha1"
+)
+
+type PromoteEnvironmentV1Alpha1Func func(
+	context.Context,
+	*connect.Request[svcv1alpha1.PromoteEnvironmentRequest],
+) (*connect.Response[svcv1alpha1.PromoteEnvironmentResponse], error)
+
+func PromoteEnvironmentV1Alpha1(
+	kc client.Client,
+) PromoteEnvironmentV1Alpha1Func {
+	return func(
+		ctx context.Context,
+		req *connect.Request[svcv1alpha1.PromoteEnvironmentRequest],
+	) (*connect.Response[svcv1alpha1.PromoteEnvironmentResponse], error) {
+		if req.Msg.GetState() == "" {
+			return nil, connect.NewError(connect.CodeInvalidArgument, errors.New("state should not be empty"))
+		}
+
+		var env kubev1alpha1.Environment
+		if err := kc.Get(ctx, client.ObjectKey{
+			Namespace: req.Msg.GetProject(),
+			Name:      req.Msg.GetName(),
+		}, &env); err != nil {
+			if kubeerr.IsNotFound(err) {
+				return nil, connect.NewError(connect.CodeNotFound,
+					fmt.Errorf("environment %q not found", req.Msg.GetName()))
+			}
+			return nil, connect.NewError(connect.CodeInternal, err)
+		}
+
+		stateExists := false
+		for _, state := range env.Status.AvailableStates {
+			if req.Msg.GetState() == state.ID {
+				stateExists = true
+				break
+			}
+		}
+		if !stateExists {
+			return nil, connect.NewError(connect.CodeNotFound,
+				fmt.Errorf("state %q not found in environment %q", req.Msg.GetState(), env.Name))
+		}
+
+		promotion := &kubev1alpha1.Promotion{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: fmt.Sprintf("%s-", req.Msg.GetName()),
```

Comment:
@krancour We discussed about naming rule in offsite meeting, but I'm still not sure if we should format promotions in same way (`<env>-to-<state_id>`). As far as I understand, the state ID should not affect the state ID hash, so I think we can use random suffix for the promotion resources 🤔 WDYT?

Reply:
@krancour can correct me if I'm wrong but to me it seemed that the `state_id` in the promotion resource name was more for convenience than for any technical reason. Perhaps we can truncate the state ID in the resource name instead?

Reply:
The name of a `Promotion` resource doesn't affect its outcome in any way. It's the only the `spec.state` field that matters.

The rationale for the naming scheme, however, is that if new state is discovered for a given `Environment` and auto-promotion is enabled, a predictable naming scheme enables the `Environment` reconciler to check whether the `Promotion` it intends to create might already exists, which is a thing that could happen if a previous attempt to reconcile the `Environment` failed late in the loop.

For this reason, the controller will have to continue using that predictable naming scheme until/unless we find a better way to ensure we never automatically create duplicate `Promotion`s.

I think using `GenerateName` here in the API server makes sense.

So... LGTM.

Aside:

> As far as I understand, the state ID should not affect the state ID hash

Just to clarify this, a state ID itself _is_ a hash. The IDs of materials that make up the state are concatenated in a canonical order and then hashed.